File: src/lua/defaults.lua
-- Built-in defaults for lsv. Loaded before user config.
--
-- config.context passed to Lua actions contains:
--   cwd                        -- current working directory
--   selected_index             -- selected row index (0-based)
--   current_len                -- number of entries in current pane
--   current_file               -- full path of highlighted item (or cwd)
--   current_file_dir           -- parent directory of highlighted item
--   current_file_name          -- basename of highlighted item
--   current_file_extension     -- extension (no dot) of highlighted item
--   current_file_ctime         -- creation time (formatted per ui.date_format)
--   current_file_mtime         -- modified time (formatted per ui.date_format)

-- Provide entries for all config values; users can override any of these.
lsv.config({
	config_version = 1,
	icons = { enabled = false, preset = nil, font = nil },
	keys = { sequence_timeout_ms = 0 },
	ui = {
		panes = { parent = 10, current = 20, preview = 70 },
		show_hidden = false,
		date_format = "%Y-%m-%d %H:%M",
		-- header format strings with placeholders
		header = {
			left = "{username}@{hostname}:{current_file}",
			right = "{current_file_size}  {owner}  {current_file_permissions}  {current_file_ctime}",
		},
		display_mode = "absolute", -- or "friendly"
		max_list_items = 5000,
		confirm_delete = true,
		-- initial listing/format defaults
		sort = "name",
		sort_reverse = false,
		show = "none",
		row = {
			icon = "{icon} ",
			left = "{name}",
			middle = "",
			right = "{info}",
		},
		-- fixed widths disabled by default (0 = auto)
		row_widths = { icon = 0, left = 0, middle = 0, right = 0 },
		-- Modal window defaults (user can override)
		modals = {
			prompt = { width_pct = 50, height_pct = 10 },
			confirm = { width_pct = 50, height_pct = 10 },
			theme = { width_pct = 60, height_pct = 60 },
		},
		-- Default dark theme; users can override fully in their config
		theme = {
			pane_bg = "#101114",
			border_fg = "gray",
			item_fg = "white",
			item_bg = nil,
			selected_item_fg = "black",
			selected_item_bg = "cyan",
			title_fg = "gray",
			title_bg = nil,
			info_fg = "gray",
			-- Per-type accents
			dir_fg = "cyan",
			dir_bg = nil,
			file_fg = "white",
			file_bg = nil,
			hidden_fg = "darkgray",
			hidden_bg = nil,
			exec_fg = "green",
			exec_bg = nil,
		},
	},
})

-- Internal quick quit
lsv.map_action("q", "Quit lsv", function(lsv, config)
	return { quit = true }
end)

-- Sort actions
lsv.map_action("sn", "Sort by name", function(lsv, config)
	config.ui.sort = "name"
end)

lsv.map_action("ss", "Sort by size", function(lsv, config)
	config.ui.sort = "size"
end)

lsv.map_action("sr", "Toggle reverse sort", function(lsv, config)
	config.ui.sort_reverse = not (config.ui.sort_reverse == true)
end)

-- Sort by modified (mtime) and created (ctime)
lsv.map_action("sm", "Sort by modified time", function(lsv, config)
	config.ui.sort = "mtime"
end)

lsv.map_action("sc", "Sort by created time", function(lsv, config)
	config.ui.sort = "created"
end)

-- Navigation (use action tables; runtime interprets nav directive)
lsv.map_action("gg", "Go to top", function(lsv, config)
	lsv.select_item(0)
end)
lsv.map_action("G", "Go to bottom", function(lsv, config)
	lsv.select_last_item()
end)

-- Info field
lsv.map_action("zn", "Info: none", function(lsv, config)
	config.ui.show = "none"
end)
lsv.map_action("zh", "Toggle Show Hidden", function(lsv, config)
	config.ui.show_hidden = not (config.ui.show_hidden == true)
end)

lsv.map_action("zs", "Info: size", function(lsv, config)
	config.ui.show = "size"
end)

lsv.map_action("zc", "Info: created date", function(lsv, config)
	config.ui.show = "created"
end)

lsv.map_action("zm", "Show messages", function(lsv, config)
	lsv.toggle_show_messages()
end)

-- Display mode
lsv.map_action("zf", "Display: friendly", function(lsv, config)
	config.ui.display_mode = "friendly"
end)

lsv.map_action("za", "Display: absolute", function(lsv, config)
	config.ui.display_mode = "absolute"
end)

-- Add file/folder action
lsv.map_action("a", "Add file/folder", function(lsv, config)
	lsv.add_entry()
end)

-- Toggle last command output panel
lsv.map_action("zo", "Show output", function(lsv, config)
	lsv.toggle_output()
end)

lsv.map_action({ "ut", "Ut" }, "UI Theme picker", function(lsv, config)
	lsv.open_theme_picker()
end)

lsv.map_action("r", "Rename selected", function(lsv, config)
	lsv.rename_item()
end)
lsv.map_action(" ", "Toggle selected", function(lsv, config)
	lsv.toggle_select()
end)
lsv.map_action("c", "Copy selected", function(lsv, config)
	lsv.copy_selection()
end)
lsv.map_action("x", "Move selected", function(lsv, config)
	lsv.move_selection()
end)
lsv.map_action("v", "Paste selected", function(lsv, config)
	lsv.paste_clipboard()
end)
lsv.map_action("u", "Clear selected", function(lsv, config)
	lsv.clear_selection()
end)
lsv.map_action("D", "Delete selected", function(lsv, config)
	lsv.delete_selected_all()
end)
-- Escape key binding (also close overlays via helper)
lsv.map_action("<Esc>", "Cancel Action / Close Popups", function(lsv, config)
	lsv.close_overlays()
end)

File: src/config/defaults.rs
use super::{
  Config,
  KeyMapping,
  UiModalConfig,
  UiModals,
  UiPanes,
  UiRowFormat,
  UiRowWidths,
  UiTheme,
};

/// Built-in default keymaps defined in Rust.
/// These mirror the previous Lua defaults and are applied before user config.
pub fn rust_default_keymaps() -> Vec<KeyMapping>
{
  vec![
    KeyMapping {
      sequence:    "q".into(),
      action:      "quit".into(),
      description: Some("Quit lsv".into()),
    },
    // Sorting
    KeyMapping {
      sequence:    "sn".into(),
      action:      "sort:name".into(),
      description: Some("Sort by name".into()),
    },
    KeyMapping {
      sequence:    "ss".into(),
      action:      "sort:size".into(),
      description: Some("Sort by size".into()),
    },
    KeyMapping {
      sequence:    "sr".into(),
      action:      "sort:reverse:toggle".into(),
      description: Some("Toggle reverse sort".into()),
    },
    KeyMapping {
      sequence:    "sm".into(),
      action:      "sort:mtime".into(),
      description: Some("Sort by modified time".into()),
    },
    KeyMapping {
      sequence:    "sc".into(),
      action:      "sort:created".into(),
      description: Some("Sort by created time".into()),
    },
    // Navigation
    KeyMapping {
      sequence:    "gg".into(),
      action:      "nav:top".into(),
      description: Some("Go to top".into()),
    },
    KeyMapping {
      sequence:    "G".into(),
      action:      "nav:bottom".into(),
      description: Some("Go to bottom".into()),
    },
    // Info/Display
    KeyMapping {
      sequence:    "zn".into(),
      action:      "show:none".into(),
      description: Some("Info: none".into()),
    },
    KeyMapping {
      sequence:    "zs".into(),
      action:      "show:size".into(),
      description: Some("Info: size".into()),
    },
    KeyMapping {
      sequence:    "zc".into(),
      action:      "show:created".into(),
      description: Some("Info: created date".into()),
    },
    KeyMapping {
      sequence:    "zf".into(),
      action:      "display:friendly".into(),
      description: Some("Display: friendly".into()),
    },
    KeyMapping {
      sequence:    "za".into(),
      action:      "display:absolute".into(),
      description: Some("Display: absolute".into()),
    },
    // Show hidden toggle and overlays
    KeyMapping {
      sequence:    "zh".into(),
      action:      "cmd:show_hidden_toggle".into(),
      description: Some("Toggle Show Hidden".into()),
    },
    KeyMapping {
      sequence:    "zm".into(),
      action:      "cmd:messages".into(),
      description: Some("Show Messages".into()),
    },
    KeyMapping {
      sequence:    "zo".into(),
      action:      "cmd:output".into(),
      description: Some("Show Output".into()),
    },
    // Find
    KeyMapping {
      sequence:    "/".into(),
      action:      "cmd:find".into(),
      description: Some("Find in current".into()),
    },
    KeyMapping {
      sequence:    "n".into(),
      action:      "cmd:next".into(),
      description: Some("Find next".into()),
    },
    KeyMapping {
      sequence:    "b".into(),
      action:      "cmd:prev".into(),
      description: Some("Find previous".into()),
    },
    // Theme picker
    KeyMapping {
      sequence:    "ut".into(),
      action:      "cmd:theme".into(),
      description: Some("UI Theme picker".into()),
    },
    KeyMapping {
      sequence:    "Ut".into(),
      action:      "cmd:theme".into(),
      description: Some("UI Theme picker".into()),
    },
    // File ops
    KeyMapping {
      sequence:    "a".into(),
      action:      "cmd:add".into(),
      description: Some("Add file/folder".into()),
    },
    KeyMapping {
      sequence:    "r".into(),
      action:      "cmd:rename".into(),
      description: Some("Rename selected".into()),
    },
    KeyMapping {
      sequence:    "D".into(),
      action:      "cmd:delete".into(),
      description: Some("Delete selected".into()),
    },
    // Selection
    KeyMapping {
      sequence:    " ".into(),
      action:      "cmd:select_toggle".into(),
      description: Some("Toggle selected".into()),
    },
    KeyMapping {
      sequence:    "u".into(),
      action:      "cmd:select_clear".into(),
      description: Some("Clear selected".into()),
    },
    // Clipboard
    KeyMapping {
      sequence:    "c".into(),
      action:      "clipboard:copy".into(),
      description: Some("Copy selected".into()),
    },
    KeyMapping {
      sequence:    "x".into(),
      action:      "clipboard:move".into(),
      description: Some("Move selected".into()),
    },
    KeyMapping {
      sequence:    "v".into(),
      action:      "clipboard:paste".into(),
      description: Some("Paste clipboard".into()),
    },
    // Overlays
    KeyMapping {
      sequence:    "<Esc>".into(),
      action:      "overlay:close".into(),
      description: Some("Close overlays".into()),
    },
  ]
}

/// Default header templates used when the user doesn't set `ui.header`.
pub const DEFAULT_HEADER_LEFT: &str = "{username}@{hostname}:{current_file}";
pub const DEFAULT_HEADER_RIGHT: &str = "{current_file_size}  {owner}  \
                                        {current_file_permissions}  \
                                        {current_file_ctime}";

/// Default modal sizes (percentages of terminal) mirrored by overlay fallbacks.
pub fn default_modals() -> UiModals
{
  UiModals {
    prompt:  UiModalConfig { width_pct: 50, height_pct: 10 },
    confirm: UiModalConfig { width_pct: 50, height_pct: 10 },
    theme:   UiModalConfig { width_pct: 60, height_pct: 60 },
  }
}

pub const DEFAULT_DATE_FORMAT: &str = "%Y-%m-%d %H:%M";

pub fn default_panes() -> UiPanes
{
  UiPanes { parent: 10, current: 20, preview: 70 }
}

pub fn default_row_widths() -> UiRowWidths
{
  UiRowWidths { icon: 0, left: 0, middle: 0, right: 0 }
}

pub fn default_theme() -> UiTheme
{
  UiTheme {
    pane_bg:               Some("#101114".into()),
    border_fg:             Some("gray".into()),
    item_fg:               Some("white".into()),
    item_bg:               Some("#101114".into()),
    selected_item_fg:      Some("black".into()),
    selected_item_bg:      Some("cyan".into()),
    title_fg:              Some("gray".into()),
    title_bg:              Some("#101114".into()),
    info_fg:               Some("gray".into()),
    dir_fg:                Some("cyan".into()),
    dir_bg:                Some("#101114".into()),
    file_fg:               Some("white".into()),
    file_bg:               Some("#101114".into()),
    hidden_fg:             Some("darkgray".into()),
    hidden_bg:             Some("#101114".into()),
    exec_fg:               Some("green".into()),
    exec_bg:               Some("#101114".into()),
    selection_bar_fg:      Some("cyan".into()),
    selection_bar_copy_fg: Some("green".into()),
    selection_bar_move_fg: Some("yellow".into()),
  }
}

/// Apply built-in defaults to any unset UI fields.
pub fn apply_config_defaults(cfg: &mut Config)
{
  // Icons default already fine
  // Keys default already fine
  let ui = &mut cfg.ui;
  if ui.panes.is_none()
  {
    ui.panes = Some(default_panes());
  }
  if ui.date_format.is_none()
  {
    ui.date_format = Some(DEFAULT_DATE_FORMAT.into());
  }
  if ui.row.is_none()
  {
    ui.row = Some(UiRowFormat::default());
  }
  if ui.row_widths.is_none()
  {
    ui.row_widths = Some(default_row_widths());
  }
  if ui.modals.is_none()
  {
    ui.modals = Some(default_modals());
  }
  if ui.theme.is_none()
  {
    ui.theme = Some(default_theme());
  }
  if ui.header_left.is_none()
  {
    ui.header_left = Some(DEFAULT_HEADER_LEFT.into());
  }
  if ui.header_right.is_none()
  {
    ui.header_right = Some(DEFAULT_HEADER_RIGHT.into());
  }
  if ui.display_mode.is_none()
  {
    ui.display_mode = Some("absolute".into());
  }
  if ui.sort.is_none()
  {
    ui.sort = Some("name".into());
  }
  if ui.sort_reverse.is_none()
  {
    ui.sort_reverse = Some(false);
  }
  if ui.show.is_none()
  {
    ui.show = Some("none".into());
  }
}
